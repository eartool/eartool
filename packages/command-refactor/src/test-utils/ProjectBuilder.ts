import type { FilePath, PackageName } from "@eartool/utils";
import * as path from "node:path";
import { Project } from "ts-morph";
import type { WorkspaceBuilder } from "./WorkspaceBuilder.js";

export class ProjectBuilder {
  #project: Project;
  #packagePath: FilePath;
  #packageName: PackageName;
  #workspace: WorkspaceBuilder;

  constructor(packageName: PackageName, packagePath: FilePath, workspace: WorkspaceBuilder) {
    this.#project = new Project({
      skipAddingFilesFromTsConfig: true,
      fileSystem: workspace.fileSystem,
      compilerOptions: {
        rootDir: "src",
      },
    });
    this.#packageName = packageName;
    this.#packagePath = packagePath;
    this.#workspace = workspace;
  }

  /**
   * @param filePath relative to the projectPath
   * @param contents
   * @returns
   */
  addFile(relativeFilePath: string, contents: string) {
    if (relativeFilePath.startsWith("/")) {
      throw new Error(`File path must be relative: ${relativeFilePath}`);
    }
    const filePath = path.resolve(this.#packagePath, relativeFilePath);
    const sf = this.#project.createSourceFile(filePath, contents);
    sf.saveSync();

    return this;
  }

  addDependency(packageName: string) {
    this.#workspace.addDependency(this.#packageName, packageName);
    return this;
  }

  get project() {
    return this.#project;
  }
}
