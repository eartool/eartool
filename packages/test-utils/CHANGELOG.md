# @eartool/test-utils

## 0.3.0

### Patch Changes

- Updated dependencies [bc9402b]
- Updated dependencies [f036955]
  - @eartool/utils@0.3.0

## 0.2.0

### Patch Changes

- @eartool/utils@0.2.0
