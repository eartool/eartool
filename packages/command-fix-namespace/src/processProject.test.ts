import { describe, expect, it } from "@jest/globals";
import { createTestLogger, createProjectForTest, formatTestTypescript } from "@eartool/test-utils";
import { processProject, type ProcessProjectOpts } from "./processProject.js";

const cases: {
  name: string;
  inputs: Record<string, string>;
  additionalRenames?: ProcessProjectOpts["additionalRenames"];
  removeNamespaces?: ProcessProjectOpts["removeNamespaces"];
  removeFauxNamespaces?: ProcessProjectOpts["removeFauxNamespaces"];
  organizeImports?: ProcessProjectOpts["organizeImports"];
}[] = [
  {
    name: "redeclare export",
    inputs: {
      "foo.tsx": `
          export namespace Foo {
            export function bar() {
              return 5;
            }

            export function baz() {
              return 5;
            }

            export type Thing = string;

            export interface OtherThing {
              what: Thing;
            }
          }
        `,

      "index.ts": `
          export {Foo} from "./foo";
      `,
    },
  },
  {
    name: "redeclare export w funky names",
    inputs: {
      "foo.tsx": `
          export namespace FooBar {
            export type ThingBar = string;

            export const doFoo = () => 5;

            export const VAR_BAR = 5;
          }
        `,

      "index.ts": `
          export {FooBar} from "./foo";
      `,
    },
  },
  {
    name: "redeclare export with ordering",
    inputs: {
      "foo.tsx": `
          export namespace Foo {
            export interface Thing {
              what: Thing;
            }
          }
        `,

      "index.ts": `
      export const before = 5;
          export {Foo} from "./foo";
          export const after = 5;
      `,
    },
  },
  {
    name: "doesnt clobber component exports",
    inputs: {
      "foo.tsx": `
          export namespace Foo {
              export interface Props { what: number }
          }

          export class Foo extends React.Component<Foo.Props> {

          }
      `,
      "index.ts": `
          export {Foo} from "./foo";
      `,
    },
  },
  {
    name: "function invoke within namespace",
    inputs: {
      "foo.tsx": `
          export namespace Foo {
            export function bar() {
              baz();
            }

            export function baz() {
              return 5;
            }
          }
          export function Foo() {}
        `,
    },
  },
  {
    name: "function invoke within namespace exclusive",
    inputs: {
      "foo.tsx": `
          export namespace Foo {
            export function bar() {
              baz();
            }

            export function baz() {
              return 5;
            }
          }
        `,
    },
  },
  {
    name: "combined types",
    inputs: {
      "foo.tsx": `
          export namespace AssociatedMapSection {
          export interface OgreProps {
            properties: Property<any>[];
          }
        
          export interface ReduxProps {
            appRealmId: RealmId;
            mapConfig: GaiaMapConfig;
          }
        
          export interface State {
            map?: MapSearchResult;
            user?: IAcmeUser;
            isLoading: boolean;
          }
        
          export type Props = OverviewObjectMinProps & OgreProps & ReduxProps;
      }`,
    },
  },
  {
    name: "works nicely with interfaces",
    // We cant break foo out in this case cause im too lazy to implement this another way.
    inputs: {
      "foo.tsx": `
          export namespace Foo {
              export interface Props { what: number }
          }

          class Foo extends React.Component<Foo.Props> {

          }
      `,
      "index.ts": `
          export {Foo} from "./foo";
      `,
    },
  },
  {
    name: "dont explode if error",
    // We cant break foo out in this case cause im too lazy to implement this another way.
    inputs: {
      "foo.ts": `
          const foo = 5;
          
          export namespace Wat {
              export const foo = 6;
          }
            `,
    },
  },
  {
    name: "simple",
    inputs: {
      "foo.ts": `
        const foo = 5;
        
        export namespace Wat {
          export const aasdf = 3;
          export const second = 5;

          export const thirdSpaced = 56;

          // Foo
          export const fourthWithComment = 555;
        }
        export function Wat(){}
        `,
    },
  },
  {
    name: "simple exclusive",
    inputs: {
      "foo.ts": `
        const foo = 5;
        
        export namespace Wat {
          export const aasdf = 3;
          export const second = 5;

          export const thirdSpaced = 56;

          // Foo
          export const fourthWithComment = 555;
        }
        `,
    },
  },
  {
    name: "rename in other file non-exclusive",
    inputs: {
      "wat.ts": `
          export namespace Wat {
            export const key = 3;
            export function f() { return 5; }

            export class Foo {}

            export type Baz = string;
          }
          export function Wat() {}
        `,

      "refWat.ts": `
          import {Wat} from "./wat";

          console.log(Wat.key);
          console.log(Wat.f());
          console.log(new Wat.Foo());
          const f: Wat.Baz = "hi";
        `,
    },
  },
  {
    name: "rename in other file exclusive",
    inputs: {
      "wat.ts": `
          export namespace Wat {
            export const key = 3;
            export function f() { return 5; }

            export class Foo {}

            export type Baz = string;
          }
        `,

      "refWat.ts": `
          import {Wat} from "./wat";

          console.log(Wat.key);
          console.log(Wat.f());
          console.log(new Wat.Foo());
          const f: Wat.Baz = "hi";
        `,
    },
  },
  {
    name: "Conflicting imports work non-exclusive",
    inputs: {
      "foo.ts": `
          export namespace Foo { export type Props = { hi: "mom" }}
          export function Foo() {}
      `,
      "bar.ts": `
          export namespace Bar { export type Props = { hi: "mom" }}
          export function Bar() {}
      `,
      "baz.ts": `
          import {Foo} from "./foo";
          import {Bar} from "./bar";
          export type Props = Foo.Props & Bar.Props;

      `,
    },
  },
  {
    name: "Conflicting imports work exclusive",
    inputs: {
      "foo.ts": `
          export namespace Foo { export type Props = { hi: "mom" }}

      `,
      "bar.ts": `
          export namespace Bar { export type Props = { hi: "mom" }}

      `,
      "baz.ts": `
          import {Foo} from "./foo";
          import {Bar} from "./bar";
          export type Props = Foo.Props & Bar.Props;

      `,
    },
  },
  {
    name: "Conflicting imports work mixed",
    inputs: {
      "foo.ts": `
          export namespace Foo { export type Props = { hi: "mom" }}
          export function Foo() {}
      `,
      "bar.ts": `
          export namespace Bar { export type Props = { hi: "mom" }}

      `,
      "baz.ts": `
          import {Foo} from "./foo";
          import {Bar} from "./bar";
          export type Props = Foo.Props & Bar.Props;

      `,
    },
  },
  {
    name: "Simple additional renames work",
    inputs: {
      "foo.ts": `
          import {Foo} from "lib";
          export type MyProps = Foo.Props;
      `,
    },
    additionalRenames: new Map([["lib", [{ from: ["Foo", "Props"], to: ["FooProps"] }]]]),
  },
  {
    name: "Generics work",
    inputs: {
      "foo.ts": `
          export namespace MapElementViewerProperties {
            export interface OwnProps<T extends MapElement> {
              mapElement: T;
              locked: boolean;
              section: RightMapPanelType;
            }
          
            export interface StoreProps {
              mapDataState: MapDataState;
            }
          
            export interface NectarProps {
              dispatch: Dispatch;
            }
          
            export type Props<T extends MapElement> = OwnProps<T> & StoreProps & NectarProps;
          }

          export function MapElementViewerProperties() {

          }
      `,
      "bar.ts": `
          import {MapElementViewerProperties} from "./foo";

          export type Foo<T extends MapElement> = MapElementViewerProperties.OwnProps<T>;
      `,
    },
  },
  {
    name: "Generics work with additional",
    inputs: {
      "foo.ts": `
          export namespace MapElementViewerProperties {
            export interface OwnProps<T extends MapElement> {
              mapElement: T;
              locked: boolean;
              section: RightMapPanelType;
            }
          
            export interface StoreProps {
              mapDataState: MapDataState;
            }
          
            export interface NectarProps {
              dispatch: Dispatch;
            }
          
            export type Props<T extends MapElement> = OwnProps<T> & StoreProps & NectarProps;
          }

          export function MapElementViewerProperties() {}
      `,
      "bar.ts": `
          import {MapElementViewerProperties} from "./foo";
          import {Bleh} from "somelib"

          export type Foo<T extends MapElement & Bleh.Bar.Other> = MapElementViewerProperties.OwnProps<T>;
      `,
    },
    additionalRenames: new Map([
      ["somelib", [{ from: ["Bleh", "Bar", "Other"], to: ["Moo", "Cow"] }]],
    ]),
  },
  {
    name: "Handles const pattern nicely",
    inputs: {
      "index.ts": `
        export {Foo} from "./foo";
      `,
      "foo.ts": `
        export const Foo = {
          // direct function
          baz() {
            return 5; 
          },

          bar: () => {
            return 6;
          }
        } as const;
      `,
      "bar.ts": `
        import {Foo} from "./foo";

        console.log(Foo.bar());
        console.log(Foo.baz());
      `,
      "alias.ts": `
        import {Foo as Bar} from "./foo";

        console.log(Bar.bar());
        console.log(Bar.baz());
      `,
    },
  },
  {
    name: "Handles namespace like faux namespace when possible",
    inputs: {
      "index.ts": `
        export {Foo} from "./foo";
      `,
      "foo.ts": `
        export namespace Foo {
          // direct function
          export function baz() {
            return 5; 
          }

          export const bar = () => {
            return 6;
          }
        } 
      `,
      "bar.ts": `
        import {Foo} from "./foo";

        console.log(Foo.bar());
        console.log(Foo.baz());
      `,
      "alias.ts": `
        import {Foo as Bar} from "./foo";

        console.log(Bar.bar());
        console.log(Bar.baz());
      `,
    },
  },
  {
    name: "Doesn't change as const that has constnats",
    inputs: {
      "index.ts": `
        export {Foo} from "./foo";
      `,
      "foo.ts": `
        export const Foo = {
          // direct function
          baz() {
            return 5; 
          },

          bar: () => {
            return 6;
          },

          FOO: 7
        } as const;
      `,
    },
  },
  {
    name: "Handles in file references to faux namespaces",
    inputs: {
      "index.ts": `
        export {Foo} from "./foo";
      `,
      "foo.ts": `
        export const Foo = {
          // direct function
          baz() {
            return 5; 
          },

          bar: () => {
            return Foo.baz();
          },

          
        } as const;
      `,
    },
  },
  {
    name: "Renames import specifiers as needed",
    inputs: {
      "foo.ts": `
        import {bar} from "lib";
        import baz from "lib2";

        export const Foo = {
          // direct function
          baz() {
            return bar() + baz(); 
          },

          bar: () => {
            return Foo.baz();
          },

          
        } as const;
      `,
    },
  },
  {
    name: "Renames star imports as needed",
    inputs: {
      "foo.ts": `
        import {other} from "other";
        import * as otherStar from "other";
        import otherDefault from "other";

        import * as bar from "lib";

        export const Foo = {
          // direct function
          baz() {
            return bar.baz() + other + otherStar.foo + otherDefault; 
          },

          bar: () => {
            return Foo.baz();
          },

          
        } as const;
      `,
    },
  },
  {
    name: "Properly handles conflict with module local variable or function",
    inputs: {
      "foo.ts": `
        function bar() { return 5; }
        const baz = 5;

        export const Foo = {
          baz() {
            return baz; 
          },

          bar: () => {
            return bar();
          },
        } as const;
      `,
    },
  },
  {
    name: "Properly handles conflict with import",
    inputs: {
      "foo.ts": `
        import {bar} from "lib";

        export const Foo = {
          baz() {
            return baz; 
          },

          bar: () => {
            return bar();
          },
        } as const;
      `,
    },
  },
  {
    name: "Multiple things in the file",
    inputs: {
      "foo.ts": `
        export namespace MetadataValue {
          export type Foo = string;
        }

        export const MetadataValue = {
          doThing() {
            return 5;
          }
        }

        export function foo() {
          return 5;
        }
      `,
      "bar.ts": `
        import {MetadataValue, foo} from "./foo";

        console.log(MetadataValue.doThing());
        console.log(foo());
      `,
    },

    removeNamespaces: false,
  },
  {
    name: "Doesn't produce extra import statements",
    inputs: {
      "foo.ts": `
        export namespace Foo {
          export function create() {}
          export function other() {}
        }

        export interface Foo {
          readonly foo: string;
        }
      `,
      "bar.ts": `
        import {Foo} from "./foo";

        doStuff(Foo.create());
        doStuff(Foo.create());
        doStuff(Foo.other());
      `,
    },
    removeNamespaces: true,
    organizeImports: false,
  },
  {
    name: "Handles local variable name collisions inside functions",
    inputs: {
      "sourceInstances.ts": `
        import { SourceInstance } from "../types/sourceInstance";

        interface FromArgs {
          readonly isFirst: boolean;
          readonly isLast: boolean;
        }

        export const SourceInstances = {
          isFirst(instance: SourceInstance) {
            return instance === SourceInstance.FIRST || instance === SourceInstance.ONLY;
          },

          isLast(instance: SourceInstance) {
            return instance === SourceInstance.LAST || instance === SourceInstance.ONLY;
          },

          from({ isFirst, isLast: reduce }: FromArgs) {
            if (isFirst) {
              return reduce ? SourceInstance.ONLY : SourceInstance.FIRST;
            } else {
              return reduce ? SourceInstance.LAST : SourceInstance.INNER;
            }
          },
          reduce(instance1: SourceInstance, instance2: SourceInstance) {
            const isFirst = SourceInstances.isFirst(instance1) && SourceInstances.isFirst(instance2);
            const isLast = SourceInstances.isLast(instance1) && SourceInstances.isLast(instance2);

            function from() { console.log(t) };
            from();
            return SourceInstances.from({ isFirst, isLast });
          },

        } as const;
      `,
    },
  },
  //Selectors as WatchedCasesSelectors,
  {
    name: "Fixes export star",
    inputs: {
      "index.ts": `
        export { Stuff } from "./stuff";
      `,
      "inbox/Inbox.ts": `
        import {Selectors as CaseSelectors} from "./cases";

        CaseSelectors.doStuff();
      `,
      "cases/index.ts": `
        export * from "./selectors";
      `,
      "cases/selectors.ts": `
        export namespace Selectors {
          export const doStuff = () => 5;
        }
      `,
    },
  },
  {
    name: "Fixes the exported type",
    inputs: {
      "index.ts": `
        export { Stuff } from "./stuff";
      `,
      "thing/index.ts": `
        import {Stuff} from "../stuff";

        Stuff.getInitialState();
      `,
      "stuff/index.ts": `
        export interface Stuff {
          hi: string;
        }
        export namespace Stuff {
          export const getInitialState = defaultMemoize((): Stuff => {
            return {
              "hi": "mom"
            };
          });
        }
      `,
    },
  },
  {
    name: "Doesnt export as type just because there is extra info",
    inputs: {
      "index.ts": `
        export { Stuff } from "./stuff";
      `,
      "thing/index.ts": `
        import {Stuff} from "../stuff";

        Stuff.getInitialState();
      `,
      "stuff/index.ts": `
        export interface Unrelated {}
        export class Stuff {
        }
        export namespace Stuff {
          export const getInitialState = defaultMemoize((): Stuff => {
            return {
              "hi": "mom"
            };
          });
        }

      `,
    },
  },
  {
    name: "Doesnt mess up indirect imports",
    inputs: {
      "index.ts": `
        import {Selectors} from "./cases";
        export const foo = Selectors.doStuff();
      `,
      "cases/index.ts": `
        export * from "./selectors";
      `,
      "cases/selectors.ts": `
        export namespace Selectors {
          export const doStuff = () => 5;
        }
      `,
    },
  },
  {
    name: "Enums rename properly",
    inputs: {
      "index.ts": `
      `,
      "dialog.ts": `
        export type DialogMode = A | B;

        export namespace DialogMode {
          export enum Type {
            A = "A",
            B = "B",
          }
        

          export function isA(dialog: DialogMode): dialog is A {
            // Fully qualified
            return dialog.type === DialogMode.Type.A;
          }
          
          export function isB(dialog: DialogMode): dialog is B {
            // Local to namespace
            return dialog.type === Type.B
          }
        }

        export interface A {
          type: DialogMode.Type.A
        }
        
        export interface B {
          type: DialogMode.Type.B
          onConfirm: () => void;
        }        
      `,
    },
  },
  {
    name: "Twins in namespace",
    inputs: {
      "index.ts": `
        export interface State {}
        export namespace State {
          export const ROOT_KEY = "state";
          export type ROOT_KEY = typeof ROOT_KEY;
        }
      `,
    },
  },
  {
    name: "reexports only alias a single time",
    inputs: {
      "index.ts": `export {Main} from "./main";`,
      "main/index.ts": `export {Main} from "./Main";`,
      "main/Main.ts": `
        export namespace Main {
          export interface Props {

          }
        }
        export class MainInternal extends React.Component {}
        export const Main = connect()(MainInternal);
      `,
    },
  },
];

describe("processProject", () => {
  it.each(cases)(
    "$name",
    async ({
      inputs,
      additionalRenames,
      removeFauxNamespaces,
      removeNamespaces,
      organizeImports,
    }) => {
      const logger = createTestLogger();
      try {
        const project = createProjectForTest(inputs);

        await processProject(
          { project, logger, packageName: "foo", packagePath: "/", packageJson: {} },
          {
            logger,
            additionalRenames,
            removeFauxNamespaces: removeFauxNamespaces ?? true,
            dryRun: false,
            organizeImports: organizeImports ?? true,
            removeNamespaces: removeNamespaces ?? true,
          }
        );

        const fs = project.getFileSystem();

        const output = project
          .getSourceFiles()
          .map((sf) => {
            const filePath = sf.getFilePath();
            return formatTestTypescript(
              `
        //

        //
        // PATH: '${filePath}'
        //
        ${fs.readFileSync(filePath)}
        `
            );
          })
          .join();

        expect(output).toMatchSnapshot();
      } finally {
        logger.flush();
      }
    }
  );

  it("records renames from root", async () => {
    const logger = createTestLogger();

    const project = createProjectForTest({
      "foo.ts": `
          export namespace Foo {
            export type Props = {};
          }
          export function Foo(){}
      `,
      "index.ts": `
          export {Foo} from "./foo";
      `,
    });

    const result = await processProject(
      { project, logger, packageName: "foo", packagePath: "/", packageJson: {} },
      {
        logger,
        removeNamespaces: true,
        removeFauxNamespaces: false,
        dryRun: false,
        organizeImports: true,
      }
    );
    expect(result.exportedRenames).toHaveLength(1);
    expect(result.exportedRenames[0]).toEqual({ from: ["Foo", "Props"], to: ["FooProps"] });
  });
});
