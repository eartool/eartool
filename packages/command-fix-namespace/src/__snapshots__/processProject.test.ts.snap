// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processProject Conflicting imports work exclusive 1`] = `
"//

//
// PATH: '/bar.ts'
//

export type Props = { hi: "mom" };
,//

//
// PATH: '/baz.ts'
//
import * as Bar from "./bar";
import * as Foo from "./foo";
export type Props = Foo.Props & Bar.Props;
,//

//
// PATH: '/foo.ts'
//

export type Props = { hi: "mom" };
"
`;

exports[`processProject Conflicting imports work mixed 1`] = `
"//

//
// PATH: '/bar.ts'
//

export type Props = { hi: "mom" };
,//

//
// PATH: '/baz.ts'
//
import * as Bar from "./bar";
import { Props as FooProps } from "./foo";
export type Props = FooProps & Bar.Props;
,//

//
// PATH: '/foo.ts'
//

export type Props = { hi: "mom" };

export function Foo() {}
"
`;

exports[`processProject Conflicting imports work non-exclusive 1`] = `
"//

//
// PATH: '/bar.ts'
//

export type Props = { hi: "mom" };

export function Bar() {}
,//

//
// PATH: '/baz.ts'
//
import { Props as BarProps } from "./bar";
import { Props as FooProps } from "./foo";
export type Props = FooProps & BarProps;
,//

//
// PATH: '/foo.ts'
//

export type Props = { hi: "mom" };

export function Foo() {}
"
`;

exports[`processProject Doesn't change as const that has constnats 1`] = `
"//

//
// PATH: '/foo.ts'
//
export const Foo = {
  // direct function
  baz() {
    return 5;
  },

  bar: () => {
    return 6;
  },

  FOO: 7,
} as const;
,//

//
// PATH: '/index.ts'
//
export { Foo } from "./foo";
"
`;

exports[`processProject Doesn't produce extra import statements 1`] = `
"//

//
// PATH: '/bar.ts'
//
import { createFoo, otherForFoo, Foo } from "./foo";

doStuff(createFoo());
doStuff(createFoo());
doStuff(otherForFoo());
,//

//
// PATH: '/foo.ts'
//

export function createFoo() {}
export function otherForFoo() {}

export interface Foo {
  readonly foo: string;
}
"
`;

exports[`processProject Doesnt mess up indirect imports 1`] = `
"//

//
// PATH: '/index.ts'
//
import { Selectors } from "./cases";
export const foo = Selectors.doStuff();
,//

//
// PATH: '/cases/index.ts'
//
export * as Selectors from "./selectors";
,//

//
// PATH: '/cases/selectors.ts'
//

export const doStuff = () => 5;
"
`;

exports[`processProject Enums rename properly 1`] = `
"//

//
// PATH: '/dialog.ts'
//
export type DialogMode = A | B;

export enum TypeForDialogMode {
  A = "A",
  B = "B",
}

export function isAForDialogMode(dialog: DialogMode): dialog is A {
  // Fully qualified
  return dialog.type === TypeForDialogMode.A;
}

export function isBForDialogMode(dialog: DialogMode): dialog is B {
  // Local to namespace
  return dialog.type === TypeForDialogMode.B;
}

export interface A {
  type: TypeForDialogMode.A;
}

export interface B {
  type: TypeForDialogMode.B;
  onConfirm: () => void;
}
,//

//
// PATH: '/index.ts'
//
"
`;

exports[`processProject Fixes export star 1`] = `
"//

//
// PATH: '/index.ts'
//
export { Stuff } from "./stuff";
,//

//
// PATH: '/cases/index.ts'
//
export * as Selectors from "./selectors";
,//

//
// PATH: '/cases/selectors.ts'
//

export const doStuff = () => 5;
,//

//
// PATH: '/inbox/Inbox.ts'
//
import { Selectors as CaseSelectors } from "./cases";

CaseSelectors.doStuff();
"
`;

exports[`processProject Fixes the exported type 1`] = `
"//

//
// PATH: '/index.ts'
//
export { getInitialStateForStuff, type Stuff } from "./stuff";
,//

//
// PATH: '/stuff/index.ts'
//
export interface Stuff {
  hi: string;
}

export const getInitialStateForStuff = defaultMemoize((): Stuff => {
  return {
    hi: "mom",
  };
});
,//

//
// PATH: '/thing/index.ts'
//
import { getInitialStateForStuff } from "../stuff";

getInitialStateForStuff();
"
`;

exports[`processProject Generics work 1`] = `
"//

//
// PATH: '/bar.ts'
//
import { OwnProps as MapElementViewerPropertiesOwnProps } from "./foo";

export type Foo<T extends MapElement> = MapElementViewerPropertiesOwnProps<T>;
,//

//
// PATH: '/foo.ts'
//

export interface OwnProps<T extends MapElement> {
  mapElement: T;
  locked: boolean;
  section: RightMapPanelType;
}

export interface StoreProps {
  mapDataState: MapDataState;
}

export interface NectarProps {
  dispatch: Dispatch;
}

export type Props<T extends MapElement> = OwnProps<T> &
  StoreProps &
  NectarProps;

export function MapElementViewerProperties() {}
"
`;

exports[`processProject Generics work with additional 1`] = `
"//

//
// PATH: '/bar.ts'
//
import { Moo } from "somelib";
import { OwnProps as MapElementViewerPropertiesOwnProps } from "./foo";

export type Foo<T extends MapElement & Moo.Cow> =
  MapElementViewerPropertiesOwnProps<T>;
,//

//
// PATH: '/foo.ts'
//

export interface OwnProps<T extends MapElement> {
  mapElement: T;
  locked: boolean;
  section: RightMapPanelType;
}

export interface StoreProps {
  mapDataState: MapDataState;
}

export interface NectarProps {
  dispatch: Dispatch;
}

export type Props<T extends MapElement> = OwnProps<T> &
  StoreProps &
  NectarProps;

export function MapElementViewerProperties() {}
"
`;

exports[`processProject Handles const pattern nicely 1`] = `
"//

//
// PATH: '/alias.ts'
//
import * as Bar from "./foo";

console.log(Bar.bar());
console.log(Bar.baz());
,//

//
// PATH: '/bar.ts'
//
import * as Foo from "./foo";

console.log(Foo.bar());
console.log(Foo.baz());
,//

//
// PATH: '/foo.ts'
//

// direct function
export function baz() {
  return 5;
}

export const bar = () => {
  return 6;
};
,//

//
// PATH: '/index.ts'
//
export * as Foo from "./foo";
"
`;

exports[`processProject Handles in file references to faux namespaces 1`] = `
"//

//
// PATH: '/foo.ts'
//

// direct function
export function baz() {
  return 5;
}

export const bar = () => {
  return baz();
};
,//

//
// PATH: '/index.ts'
//
export * as Foo from "./foo";
"
`;

exports[`processProject Handles local variable name collisions inside functions 1`] = `
"//

//
// PATH: '/sourceInstances.ts'
//
import { SourceInstance } from "../types/sourceInstance";

interface FromArgs {
  readonly isFirst: boolean;
  readonly isLast: boolean;
}

export function isFirst(instance: SourceInstance) {
  return instance === SourceInstance.FIRST || instance === SourceInstance.ONLY;
}

export function isLast(instance: SourceInstance) {
  return instance === SourceInstance.LAST || instance === SourceInstance.ONLY;
}

export function from({ isFirst: isFirst0, isLast: reduce0 }: FromArgs) {
  if (isFirst0) {
    return reduce0 ? SourceInstance.ONLY : SourceInstance.FIRST;
  } else {
    return reduce0 ? SourceInstance.LAST : SourceInstance.INNER;
  }
}
export function reduce(instance1: SourceInstance, instance2: SourceInstance) {
  const isFirst0 = isFirst(instance1) && isFirst(instance2);
  const isLast0 = isLast(instance1) && isLast(instance2);

  function from0() {
    console.log(t);
  }
  from0();
  return from({ isFirst: isFirst0, isLast: isLast0 });
}
"
`;

exports[`processProject Handles namespace like faux namespace when possible 1`] = `
"//

//
// PATH: '/alias.ts'
//
import * as Bar from "./foo";

console.log(Bar.bar());
console.log(Bar.baz());
,//

//
// PATH: '/bar.ts'
//
import * as Foo from "./foo";

console.log(Foo.bar());
console.log(Foo.baz());
,//

//
// PATH: '/foo.ts'
//

// direct function
export function baz() {
  return 5;
}

export const bar = () => {
  return 6;
};
,//

//
// PATH: '/index.ts'
//
export * as Foo from "./foo";
"
`;

exports[`processProject Multiple things in the file 1`] = `
"//

//
// PATH: '/bar.ts'
//
import { MetadataValue, foo } from "./foo";

console.log(MetadataValue.doThing());
console.log(foo());
,//

//
// PATH: '/foo.ts'
//
export namespace MetadataValue {
  export type Foo = string;
}

export const MetadataValue = {
  doThing() {
    return 5;
  },
};

export function foo() {
  return 5;
}
"
`;

exports[`processProject Properly handles conflict with import 1`] = `
"//

//
// PATH: '/foo.ts'
//
import { bar as bar0 } from "lib";

export function baz() {
  return baz;
}

export const bar = () => {
  return bar0();
};
"
`;

exports[`processProject Properly handles conflict with module local variable or function 1`] = `
"//

//
// PATH: '/foo.ts'
//
function bar0() {
  return 5;
}
const baz0 = 5;

export function baz() {
  return baz0;
}

export const bar = () => {
  return bar0();
};
"
`;

exports[`processProject Renames import specifiers as needed 1`] = `
"//

//
// PATH: '/foo.ts'
//
import { bar as bar0 } from "lib";
import baz0 from "lib2";

// direct function
export function baz() {
  return bar0() + baz0();
}

export const bar = () => {
  return baz();
};
"
`;

exports[`processProject Renames star imports as needed 1`] = `
"//

//
// PATH: '/foo.ts'
//
import * as otherStar from "other";
import otherDefault, { other } from "other";

import * as bar0 from "lib";

// direct function
export function baz() {
  return bar0.baz() + other + otherStar.foo + otherDefault;
}

export const bar = () => {
  return baz();
};
"
`;

exports[`processProject Simple additional renames work 1`] = `
"//

//
// PATH: '/foo.ts'
//
import { FooProps } from "lib";
export type MyProps = FooProps;
"
`;

exports[`processProject Twins in namespace 1`] = `
"//

//
// PATH: '/index.ts'
//
export interface State {}

export const STATE_ROOT_KEY = "state";
export type STATE_ROOT_KEY = typeof STATE_ROOT_KEY;
"
`;

exports[`processProject combined types 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export interface OgreProps {
  properties: Property<any>[];
}

export interface ReduxProps {
  appRealmId: RealmId;
  mapConfig: GaiaMapConfig;
}

export interface State {
  map?: MapSearchResult;
  user?: IAcmeUser;
  isLoading: boolean;
}

export type Props = OverviewObjectMinProps & OgreProps & ReduxProps;
"
`;

exports[`processProject doesnt clobber component exports 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export interface Props {
  what: number;
}

export class Foo extends React.Component<Props> {}
,//

//
// PATH: '/index.ts'
//
export { Foo, type Props as FooProps } from "./foo";
"
`;

exports[`processProject dont explode if error 1`] = `
"//

//
// PATH: '/foo.ts'
//
const foo0 = 5;

export const foo = 6;
"
`;

exports[`processProject function invoke within namespace 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export function barForFoo() {
  bazForFoo();
}

export function bazForFoo() {
  return 5;
}

export function Foo() {}
"
`;

exports[`processProject function invoke within namespace exclusive 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export function bar() {
  baz();
}

export function baz() {
  return 5;
}
"
`;

exports[`processProject redeclare export 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export function bar() {
  return 5;
}

export function baz() {
  return 5;
}

export type Thing = string;

export interface OtherThing {
  what: Thing;
}
,//

//
// PATH: '/index.ts'
//
export * as Foo from "./foo";
"
`;

exports[`processProject redeclare export w funky names 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export type ThingBar = string;

export const doFoo = () => 5;

export const VAR_BAR = 5;
,//

//
// PATH: '/index.ts'
//
export * as FooBar from "./foo";
"
`;

exports[`processProject redeclare export with ordering 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export interface Thing {
  what: Thing;
}
,//

//
// PATH: '/index.ts'
//
export const before = 5;
export * as Foo from "./foo";
export const after = 5;
"
`;

exports[`processProject rename in other file exclusive 1`] = `
"//

//
// PATH: '/refWat.ts'
//
import * as Wat from "./wat";

console.log(Wat.key);
console.log(Wat.f());
console.log(new Wat.Foo());
const f: Wat.Baz = "hi";
,//

//
// PATH: '/wat.ts'
//

export const key = 3;
export function f() {
  return 5;
}

export class Foo {}

export type Baz = string;
"
`;

exports[`processProject rename in other file non-exclusive 1`] = `
"//

//
// PATH: '/refWat.ts'
//
import { BazForWat, FooForWat, fForWat, keyForWat } from "./wat";

console.log(keyForWat);
console.log(fForWat());
console.log(new FooForWat());
const f: BazForWat = "hi";
,//

//
// PATH: '/wat.ts'
//

export const keyForWat = 3;
export function fForWat() {
  return 5;
}

export class FooForWat {}

export type BazForWat = string;

export function Wat() {}
"
`;

exports[`processProject simple 1`] = `
"//

//
// PATH: '/foo.ts'
//
const foo = 5;

export const aasdfForWat = 3;
export const secondForWat = 5;

export const thirdSpacedForWat = 56;

// Foo
export const fourthWithCommentForWat = 555;

export function Wat() {}
"
`;

exports[`processProject simple exclusive 1`] = `
"//

//
// PATH: '/foo.ts'
//
const foo = 5;

export const aasdf = 3;
export const second = 5;

export const thirdSpaced = 56;

// Foo
export const fourthWithComment = 555;
"
`;

exports[`processProject works nicely with interfaces 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export interface Props {
  what: number;
}

class Foo extends React.Component<Foo.Props> {}
,//

//
// PATH: '/index.ts'
//
export * as Foo from "./foo";
"
`;
